src/utils/single-implementation.js

```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+/* global WeakMap, Map, Set */
+
+const UNIQUE_ADDON_MAP = new WeakMap();
+
+/**
+ * Traverse the project's addons tree to determine singleton root,
+ * cache the boolean result in project x addonName matrix
+ */
+module.exports.hasSingleImplementation = hasSingleImplementation;
+function hasSingleImplementation(name, project) {
+  if (!UNIQUE_ADDON_MAP.has(project)) {
+    UNIQUE_ADDON_MAP.set(project, new Map());
+  }
+  let map = UNIQUE_ADDON_MAP.get(project);
+  if (map.has(name)) {
+    return map.get(name);
+  }
+
+  let isSingleton = dfsRecursive(project, name);
+  map.set(name, isSingleton);
+  return isSingleton;
+}
+
+/**
+ * Traverse the addons tree, early return if duplication is found
+ */
+module.exports.dfsRecursive = dfsRecursive;
+function dfsRecursive(current, name, addonsRootSet = new Set(), addons) {
+  let currentAddons = current.addons || [];
+  for (let addon of currentAddons) {
+    if (addon.name === name) {
+      if (Array.isArray(addons) && !addonsRootSet.has(addon.root)) {
```

Comment:
I don't understand this conditional `findAddonsByName` should not "uniqueify" the addons, it should be the full list. I would expect to always push (regardless of the set contents).

Reply:
I don't think we expect to see the same implementation pushed twice.

Reply:
I disagree. For this to work for the ember-test-waiters case, we need the addon instances of each time we find this addon in the addon graph (so that we can mutate and make it a no op).

Reply:
Ah that is a different use-case, I did not think about. Yup that makes sense.